### PR TITLE
Mentor System + Who

### PR DESCRIPTION
**Pull request by j-mie**
Ports over Nalar's Who
Ports over my Mentor system
Ports over changes made to the system such as memo
Fixes a couple of bugs in the mentor system too

### CODE REVIEW

File: code/modules/hippie/mentor/follow.dm [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	set name = "Stop Following"
+	set desc = "Stop following the followed."
+
+	if(!check_mentor())
+		return
+
+	usr.reset_perspective(null)
+	src.verbs -= /client/proc/mentor_unfollow
+
+	var/following = null
```
Comment (Carbonhell) [inline code/modules/hippie/mentor/follow.dm:38]: Why are you using this var? You're not resetting mentor.following either, is that what you actually want?

File: code/modules/hippie/mentor/holder2.dm [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		owner = C
+		mentors |= C
+
+/datum/mentors/proc/disassociate()
+	if(owner)
+		mentors -= owner
+		owner = null
+
+/client/proc/dementor()
+	mentor_datums -= ckey
```
Comment (Carbonhell) [inline code/modules/hippie/mentor/holder2.dm]: shouldn't you also delete the mentors datum?

File: code/modules/hippie/mentor/verbs/mentor_memo.dm [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	if(!check_rights(0))	return
+	if(!dbcon.IsConnected())
+		src << "<span class='danger'>Failed to establish database connection.</span>"
+		return
+	var/memotask = input(usr,"Choose task.","Memo") in list("Show","Write","Edit","Remove")
+	if(!memotask)
+		return
+	mentor_memo_output(memotask)
+
+/client/proc/show_mentor_memo()
```
Comment (Carbonhell) [inline code/modules/hippie/mentor/verbs/mentor_memo.dm:13]: ...isn't this redundant?

File: code/modules/hippie/mentor/verbs/mentorhelp.dm [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/client/verb/mentorhelp(msg as text)
+	set category = "Mentor"
+	set name = "Mentorhelp"
+
+	//remove out adminhelp verb temporarily to prevent spamming of mentors.
```
Comment (Carbonhell) [inline code/modules/hippie/mentor/verbs/mentorhelp.dm]: mentorhelp rather

File: code/modules/hippie/mentor/verbs/mentorpm.dm [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	var/target = input(src,"To whom shall we send a message?","Mentor PM",null) in sorted|null
+	cmd_mentor_pm(targets[target],null)
+	feedback_add_details("Mentor_verb","APM") //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
+
+
+//takes input from cmd_mentor_pm_context, cmd_Mentor_pm_panel or /client/Topic and sends them a PM.
+//Fetching a message if needed. src is the sender and C is the target client
+/client/proc/cmd_mentor_pm(whom, msg)
+	var/client/C
+	if(istype(whom,/mob))
```
Comment (Carbonhell) [inline code/modules/hippie/mentor/verbs/mentorpm.dm]: could just do ismob

**Review by Carbonhell — CHANGES_REQUESTED** [review]

File: code/modules/hippie/mentor/verbs/mentor_memo.dm (hunk 3 above)
Comment (j-mie) [reply code/modules/hippie/mentor/verbs/mentor_memo.dm:13]: Why?

**Review by Carbonhell — APPROVED** [review]

**Pull request merged**